glean-core/src/scheduler.rs

```diff
@@ -0,0 +1,496 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! # Metrics Ping Scheduler
+//!
+//! The Metrics Ping Scheduler (MPS) is responsible for scheduling "metrics" pings.
+//! It implements the spec described in
+//! [the docs](https://mozilla.github.io/glean/book/user/pings/metrics.html#scheduling)
+
+use crate::metrics::{DatetimeMetric, StringMetric, TimeUnit};
+use crate::{local_now_with_offset, CommonMetricData, Glean, Lifetime, INTERNAL_STORAGE};
+use chrono::prelude::*;
+use chrono::Duration;
+use once_cell::sync::Lazy;
+use std::sync::{Arc, Condvar, Mutex};
+use std::thread::JoinHandle;
+
+const SCHEDULED_HOUR: u32 = 4;
+
+// Clippy thinks an AtomicBool would be preferred, but Condvar requires a full Mutex.
+// See https://github.com/rust-lang/rust-clippy/issues/1516
+#[allow(clippy::mutex_atomic)]
+static TASK_CONDVAR: Lazy<Arc<(Mutex<bool>, Condvar)>> =
+    Lazy::new(|| Arc::new((Mutex::new(false), Condvar::new())));
+
+trait MetricsPingSubmitter {
+    fn submit_metrics_ping(&self, glean: &Glean, reason: Option<&str>, now: DateTime<FixedOffset>);
+}
+trait MetricsPingScheduler {
+    fn start_scheduler(
+        &self,
+        submitter: impl MetricsPingSubmitter + Send + 'static,
+        now: DateTime<FixedOffset>,
+        when: When,
+    );
+}
+
+struct GleanMetricsPingSubmitter {}
+impl MetricsPingSubmitter for GleanMetricsPingSubmitter {
+    fn submit_metrics_ping(&self, glean: &Glean, reason: Option<&str>, now: DateTime<FixedOffset>) {
+        if let Err(err) = glean.submit_ping_by_name("metrics", reason) {
+            log::warn!(
+                "Error submitting 'metrics' ping with reason {:?}: {}",
+                reason,
+                err
+            );
+        }
+        // Always update the collection date, even if the ping submission errored.
+        get_last_sent_time_metric().set(&glean, Some(now));
+    }
+}
+struct GleanMetricsPingScheduler {}
+impl MetricsPingScheduler for GleanMetricsPingScheduler {
+    fn start_scheduler(
+        &self,
+        submitter: impl MetricsPingSubmitter + Send + 'static,
+        now: DateTime<FixedOffset>,
+        when: When,
+    ) {
+        start_scheduler(submitter, now, when);
+    }
+}
+
+/// Performs startup checks to decide when to schedule the next "metrics" ping collection.
+/// Must be called before draining the preinit queue.
```

Comment:
Maybe make this more obvious then?

```suggestion
/// **Must** be called before draining the preinit queue.
```